CLDSRV-167: implement put bucket tagging

KillianG opened the pull request:
Checking order has been changed to fit with AWS one, meaning that we
raise the first error we face

- Checking every tag one by one
- Checking that the tag key is unique, the error is raised only if we
  already seen a tag with the same key before
- Checking that the key length is less than 128
- Checking that the tag value length is less than 256
- After seeing more than 50 tags, if there is one more, raise the too
  many tags error

Commit:
CLDSRV-167: implement put bucket tagging

bert-e commented:
# Hello killiang,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/cloudserver/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-167 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `7.10.4`

* `8.2.21`

* `8.3.13`

* `8.4.9`


Please check the `Fix Version/s` of CLDSRV-167, or the target
branch of this pull request.

XinLi0207 reviewed (commented):
LGTM, just a small question: Is there a max number of tags that can be added to a bucket or it's unlimited?

williamlardier commented on lib/api/bucketPutTagging.js:
```diff
@@ -0,0 +1,104 @@
+            (err, bucket) => next(err, { result, bucket })),
+        ({ result, bucket }, next) => checkTags(result.Tagging.TagSet.Tag, bucket, log, next),
+        ({ tags, bucket }, next) => {
+            bucket.setTags(tags);
```
Random question, the only place I found this function was in ObjectMD (in Arsenal), but the function overwrite all the tags with the previously-set tags. In your tests, can you try to put two tags (one by one) and check, after the two PutBucketTagging, that both tags are in the bucket MD?

williamlardier replied:
Didn't see the api note: "When this operation sets the tags for a bucket, it will overwrite any current tags the bucket already has. You cannot use this operation to add tags to an existing list of tags.", so all good

ghost commented:
@XinLiScality yes, according to https://docs.aws.amazon.com/AmazonS3/latest/userguide/tagging-managing.html, it's 10. Do we have the same limits in the requirements?
Edit: maybe it's only for object tagging?

XinLi0207 commented on tests/functional/aws-node-sdk/test/bucket/putBucketTagging.js:
```diff
@@ -0,0 +1,177 @@
+
+    afterEach(done => s3.deleteBucket({ Bucket: bucket }, done));
+
+    it('should not add tag if tagKey not unique', done => {
```
do you think it's worth testing if a bucket tagging can overwrite the one with the same key?

WillToozsScality replied:
It does this by default, all tags are overwritten every time a tagset is uploaded

williamlardier commented on lib/api/bucketPutTagging.js:
```diff
@@ -0,0 +1,104 @@
+            metadata.updateBucket(bucket.getName(), bucket, log, err =>
+                next(err, bucket));
+        },
+    ], (err, bucket) => {
+        const corsHeaders = collectCorsHeaders(request.headers.origin,
+            request.method, bucket);
+        return callback(err, corsHeaders);
+    });
+}
```
```suggestion
    ], (err, bucket) => {
        const corsHeaders = collectCorsHeaders(request.headers.origin,
            request.method, bucket);
        pushMetric('putBucketTagging', log, {
            authInfo,
            bucket: bucketName,
        });
        return callback(err, corsHeaders);
    });
}
```

WillToozsScality replied:
done dee6e21fdf6c2fef04ad5784060fb2be8d5ef149

williamlardier reviewed (commented):
Suggest adding the pushMetric call in the API

anurag4DSB commented on lib/api/bucketPutTagging.js:
```diff
@@ -0,0 +1,104 @@
+    if (unique.length !== tags.length) {
+        errMsg = tagKeyNotUnique;
+    }
+    tags.forEach(item => {
+        if (item.Key.length > 128) {
+            errMsg = tagKeyInvalid;
+        }
+        if (item.Value.length > 256) {
+            errMsg = tagValueInvalid;
+        }
+    });
```
As per this we check all the tagsets and return the last known error.
In AWS, I tried the same, it returns on the first error that was found.


For example, I tried the below JSON:
```json
{
    "TagSet": [
      {
        "Key": "organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1",
        "Value": "marketing"
      },
      {
        "Key": "o121",
        "Value": "rganization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1"
      }
    ]
 }
```

Cloudserver (current PR code) gives me invalid value:
```sh
❯ aws s3api put-bucket-tagging --bucket test1 --tagging file://tagset.json --profile vault --endpoint-url http://localhost:8000

An error occurred (InvalidTag) when calling the PutBucketTagging operation: The TagValue you have provided is invalid
```

Where as AWS gives me invalid Tag Key:
```sh
❯ aws s3api put-bucket-tagging --bucket anuragplatform --tagging file://tagset.json

An error occurred (InvalidTag) when calling the PutBucketTagging operation: The TagKey you have provided is invalid
```

WillToozsScality replied:
See comment here for how I fixed error prioritisation https://github.com/scality/cloudserver/pull/4451#issuecomment-1108687268

anurag4DSB reviewed (changes requested):
I also tried uploading a TagSet of more than 50 tags, AWS sent me back an error, but cloudserver did not
```
❯ aws s3api put-bucket-tagging --bucket anuragplatform --tagging file://tagset.json

An error occurred (BadRequest) when calling the PutBucketTagging operation: Bucket tag count cannot be greater than 50
❯ aws s3api put-bucket-tagging --bucket test1 --tagging file://tagset.json --profile vault --endpoint-url http://localhost:8000
```
We should adhere to the restrictions mentioned in requirements: https://github.com/scality/citadel/blob/development/1.0/docs/requirements/bucket-tagging-s3c.md#restrictions-on-tagging

WillToozsScality commented on lib/api/bucketPutTagging.js:
```diff
@@ -0,0 +1,104 @@
+    let errMsg = null;
+    if (unique.length !== tags.length) {
+        errMsg = tagKeyNotUnique;
+    }
```
you're right, done dee6e21fdf6c2fef04ad5784060fb2be8d5ef149

williamlardier commented on lib/api/bucketPutTagging.js:
```diff
@@ -0,0 +1,121 @@
+const tagKeyNotUnique = 'Cannot provide multiple Tags with the same key';
+const tagKeyInvalid = 'The TagKey you have provided is invalid';
+const tagValueInvalid = 'The TagValue you have provided is invalid';
+const tooManyTags = 'Bucket tag count cannot be greater than 50'
```
Otherwise LGTM (maybe the `50` could be stored though)

WillToozsScality replied:
Should've done a lint before, apologies. Will fix in next push

WillToozsScality replied:
done 090e8a290b6adf2d69bce53dc98eefe539751b33

ghost commented:
As per what I've tested with AWS, error prioritization is now as I've observed:
- failing on the first error it comes across as it goes through through tags one by one (including unique keys: a dupe key-pair could be at positions 1 and 49, but the error received would be TagValue invalid if an invalid tag is at position 48)
- THEN checking tag count once it's evaluated all the tags (no matter array size, as inefficient as that is...)

So as much as I hate pushing such inefficient code, here we are 090e8a290b6adf2d69bce53dc98eefe539751b33

Please do make any suggestions to optimize this !

anurag4DSB commented:
> As per what I've tested with AWS, error prioritization is now as I've observed:
> 
> * failing on the first error it comes across as it goes through through tags one by one (including unique keys: a dupe key-pair could be at positions 1 and 49, but the error received would be TagValue invalid if an invalid tag is at position 48)
> * THEN checking tag count once it's evaluated all the tags (no matter array size, as inefficient as that is...)
> 
> So as much as I hate pushing such inefficient code, here we are [090e8a2](https://github.com/scality/cloudserver/commit/090e8a290b6adf2d69bce53dc98eefe539751b33)
> 
> Please do make any suggestions to optimize this !

@WillToozsScality Do you see similar for AWS-SDK as well. There is a possibility that AWS CLI is doing something than what just BOTO3 does. 
Also, as per the current code we will go through all possible tags, which is expensive ofcourse like you mentioned.
I tried AWS and they do have some kind of rate limiting post 3500 hundred tags(depends on the content of key-value).

Either we figure out what AWS does for rate-limiting(more below, might need some sort of design if not already present), or we put the condition of checking the number of tags first(error out if more than 50) and move the rest of the checking logic after that.


AWS behavior with 3500+ tags:

```
❯ aws s3api put-bucket-tagging --bucket anuragplatform --tagging file://tagset2.json

An error occurred (MalformedXML) when calling the PutBucketTagging operation: The XML you provided was not well-formed or did not validate against our published schema
```

TagSet used:
```sh
{
    "TagSet":
	[{"Key":"a0","Value":"a0"},{"Key":"a1","Value":"a1"},{"Key":"a2","Value":"a2"},{"Key":"a3","Value":"a3"},{"Key":"a4","Value":"a4"},{"Key":"a5","Value":"a5"},{"Key":"a6","Value":"a6"},{"Key":"a7","Value":"a7"},{"Key":"a8","Value":"a8"},{"Key":"a9","Value":"a9"},{"Key":"a10","Value":"a10"},
.
.
.
{"Key":"a3545","Value":"a3545"},{"Key":"a3546","Value":"a3546"},{"Key":"a3547","Value":"a3547"},{"Key":"a3548","Value":"a3548"},{"Key":"a3549","Value":"a3549"},{"Key":"a3550","Value":"a3550"},{"Key":"a3551","Value":"a3551"},{"Key":"a3552","Value":"a3552"},{"Key":"a3553","Value":"a3553"},{"Key":"a3554","Value":"a3554"},{"Key":"a3555","Value":"a3555"},{"Key":"a3556","Value":"a3556"},{"Key":"a3557","Value":"a3557"},{"Key":"a3558","Value":"a3558"},{"Key":"a3559","Value":"a3559"},{"Key":"a3560","Value":"a3560"}]
}
```

For the same tagSet we check everything:
```
❯ aws s3api put-bucket-tagging --bucket test1 --tagging file://tagset2.json --endpoint-url http://localhost:8000 --profile vault

An error occurred (BadRequest) when calling the PutBucketTagging operation: Bucket tag count cannot be greater than 50
```

anurag4DSB reviewed (changes requested):
Requesting for last bit of changes.
Would recommend, we put the condition of checking the number of tags first(error out if more than 50) and move the rest of the checking logic after that.
I know this is not what AWS does, but post a certain value, they do some rate limiting.

Lastly, AWSCLI provides this option: `[--expected-bucket-owner <value>]`
Do we support that? If not then we need to do that as well.

Commit:
CLDSRV-167: Implement bucket owner check and change check order

Checking order has been changed to fit with AWS one, meaning that we
raise the first error we face

- Checking every tag one by one
- Checking that the tag key is unique, the error is raised only if we
  already seen a tag with the same key before
- Checking that the key length is less than 128
- Checking that the tag value length is less than 256
- After seeing more than 50 tags, if there is one more, raise the too
  many tags error

KillianG commented:
> Requesting for last bit of changes. Would recommend, we put the condition of checking the number of tags first(error out if more than 50) and move the rest of the checking logic after that. I know this is not what AWS does, but post a certain value, they do some rate limiting.

This is done now [here](https://github.com/scality/cloudserver/pull/4451/commits/aab111506a9d3e68df36cf9c058b8301f24fc960)

> Lastly, AWSCLI provides this option: `[--expected-bucket-owner <value>]` Do we support that? If not then we need to do that as well.

We do support that now [here](https://github.com/scality/cloudserver/pull/4451/commits/b29756602ba8c5671672f8b0b4d1c25d34cc80c4)

WillToozsScality commented on lib/api/bucketPutTagging.js:
```diff
@@ -58,6 +58,16 @@ function checkTags(tagSet, bucket, log, cb) {
+            const error = errors.BadRequest.customizeDescription(
+                tooManyTags);
+            return cb(error);
+        }
```
While in the end this will still be an improvement, when Anurag asked to put it before the other tests, he likely meant before the loop. Either way I'm sure it's not a big deal at this point.

KillianG replied:
I checked, and AWS raises the first error it faces, so if I check for 50 tags before anything else we might end up showing "error more than 50 tags" before showing "invalid tag".

By doing this way, we check all errors on the first 50 tags and if there is no errors, but there is more than 50 tags, we raise the "more than 50 tags error" without going further than 50, so if there is 999.999.999.999 tags, and they are all valid, we only check the first 50

anurag4DSB commented on lib/api/bucketPutTagging.js:
```diff
@@ -0,0 +1,129 @@
+    });
+}
+
+function checkTags(tagSet, bucket, log, cb) {
+    if (tagSet === undefined) {
+        return cb(null, []);
+    }
+    const tags = Array.isArray(tagSet) ? tagSet : [tagSet];
+    let errMsg = null;
+    const uniqueKeys = [];
+    for (let i = 0; i < tags.length && !errMsg; i++) {
+        if (uniqueKeys.find(key => tags[i].Key === key) !== undefined) {
+            errMsg = tagKeyNotUnique;
+        } else {
+            uniqueKeys.push(tags[i].Key);
+        }
+        if (tags[i].Key.length > 128) {
+            errMsg = tagKeyInvalid;
+        } else if (tags[i].Value.length > 256) {
+            errMsg = tagValueInvalid;
+        }
+        if (i > 50) {
+            log.debug(tooManyTags,
+                {
+                    method: 'bucketPutTagging',
+                    error: errors.BadRequest,
+                });
+            const error = errors.BadRequest.customizeDescription(
+                tooManyTags);
+            return cb(error);
+        }
```
This logic deserves an in-code comment.
And a commit message story(while squashing commits)
A small example: https://github.com/scality/Vault/commit/7c73cc53b673665344329bd83ff5633f29c6e138
But you can even make bullet points. The only requirement is the first line should comply with bert-e requirements.

anurag4DSB replied:
feel free to squash for the next review :)

KillianG replied:
Squashed and fixed [here](https://github.com/scality/cloudserver/pull/4451/commits/ea107a50e15842ac0a046b10bcb55ff4bc2e773d)

anurag4DSB reviewed (changes requested):
the code LGTM, but we need more in code comments and a bit more detailed story in the commit message.

anurag4DSB reviewed (approved):
LGTM, I would still say deserves a in-code comment :) 
Specially where we do not comply with AWS

KillianG commented:
> LGTM, I would still say deserves a in-code comment :) Specially where we do not comply with AWS

AFAIK, we totally comply with AWS.. @WillToozsScality can you confirm ?

Commit:
Merge branch 'development/7.10' into improvement/CLDSRV-167-put-bucket-tagging

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-167 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `7.10.5`

* `8.4.11`

* `8.5.1`


Please check the `Fix Version/s` of CLDSRV-167, or the target
branch of this pull request.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.4/improvement/CLDSRV-167-put-bucket-tagging` with contents from `improvement/CLDSRV-167-put-bucket-tagging`
and `development/8.4`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.4/improvement/CLDSRV-167-put-bucket-tagging origin/development/8.4
 $ git merge origin/improvement/CLDSRV-167-put-bucket-tagging
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.4/improvement/CLDSRV-167-put-bucket-tagging
```

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/CLDSRV-167-put-bucket-tagging` into
`development/7.10`
* [w/8.4/improvement/CLDSRV-167-put-bucket-tagging](https://github.com/scality/cloudserver/compare/development/8.4...w/8.4/improvement/CLDSRV-167-put-bucket-tagging) will be merged into `development/8.4` (pull request #4566)
* [w/8.5/improvement/CLDSRV-167-put-bucket-tagging](https://github.com/scality/cloudserver/compare/development/8.5...w/8.5/improvement/CLDSRV-167-put-bucket-tagging) will be merged into `development/8.5` (pull request #4567)



The following branches will **NOT** be impacted:

* `development/6.4`
* `development/7.4`
* `stabilization/7.10.4`




*Follow* integration pull requests if you would like to be notified of
build statuses by email.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

KillianG commented:
/reset

bert-e commented:
# Lossy reset warning



There seems to be *manual* commits on integration branches (e.g. conflict
resolutions) that will be lost if you chose to *reset*.

**You can use the `force_reset` command if you still want me
to delete those branches.**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

KillianG commented:
/force_reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.4/improvement/CLDSRV-167-put-bucket-tagging` with contents from `improvement/CLDSRV-167-put-bucket-tagging`
and `development/8.4`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.4/improvement/CLDSRV-167-put-bucket-tagging origin/development/8.4
 $ git merge origin/improvement/CLDSRV-167-put-bucket-tagging
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.4/improvement/CLDSRV-167-put-bucket-tagging
```

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.5/improvement/CLDSRV-167-put-bucket-tagging` with contents from `w/8.4/improvement/CLDSRV-167-put-bucket-tagging`
and `development/8.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.5/improvement/CLDSRV-167-put-bucket-tagging origin/development/8.5
 $ git merge origin/w/8.4/improvement/CLDSRV-167-put-bucket-tagging
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.5/improvement/CLDSRV-167-put-bucket-tagging
```

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/CLDSRV-167-put-bucket-tagging` into
`development/7.10`
* [w/8.4/improvement/CLDSRV-167-put-bucket-tagging](https://github.com/scality/cloudserver/compare/development/8.4...w/8.4/improvement/CLDSRV-167-put-bucket-tagging) will be merged into `development/8.4` (pull request #4568)
* [w/8.5/improvement/CLDSRV-167-put-bucket-tagging](https://github.com/scality/cloudserver/compare/development/8.5...w/8.5/improvement/CLDSRV-167-put-bucket-tagging) will be merged into `development/8.5` (pull request #4569)



The following branches will **NOT** be impacted:

* `development/6.4`
* `development/7.4`
* `stabilization/7.10.4`




*Follow* integration pull requests if you would like to be notified of
build statuses by email.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

KillianG commented:
/force_reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.4/improvement/CLDSRV-167-put-bucket-tagging` with contents from `improvement/CLDSRV-167-put-bucket-tagging`
and `development/8.4`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.4/improvement/CLDSRV-167-put-bucket-tagging origin/development/8.4
 $ git merge origin/improvement/CLDSRV-167-put-bucket-tagging
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.4/improvement/CLDSRV-167-put-bucket-tagging
```

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/CLDSRV-167-put-bucket-tagging` into
`development/7.10`
* [w/8.4/improvement/CLDSRV-167-put-bucket-tagging](https://github.com/scality/cloudserver/compare/development/8.4...w/8.4/improvement/CLDSRV-167-put-bucket-tagging) will be merged into `development/8.4` (pull request #4570)
* [w/8.5/improvement/CLDSRV-167-put-bucket-tagging](https://github.com/scality/cloudserver/compare/development/8.5...w/8.5/improvement/CLDSRV-167-put-bucket-tagging) will be merged into `development/8.5` (pull request #4571)



The following branches will **NOT** be impacted:

* `development/6.4`
* `development/7.4`
* `stabilization/7.10.4`




*Follow* integration pull requests if you would like to be notified of
build statuses by email.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

Commit:
Update bucketPutTagging.js

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-167 contains:


* `7.10.5`

* `8.4.12`

* `8.5.1`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `7.10.4`

* `8.4.12`

* `8.5.3`


Please check the `Fix Version/s` of CLDSRV-167, or the target
branch of this pull request.

bert-e commented:
# Conflict




A conflict has been raised during the update of
integration branch `w/8.4/improvement/CLDSRV-167-put-bucket-tagging` with contents from `improvement/CLDSRV-167-put-bucket-tagging`
and `development/8.4`.

Please resolve the conflict on **the integration branch** (`w/8.4/improvement/CLDSRV-167-put-bucket-tagging`).


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout w/8.4/improvement/CLDSRV-167-put-bucket-tagging
 $ git pull  # or "git reset --hard origin/w/8.4/improvement/CLDSRV-167-put-bucket-tagging"
 $ git merge origin/development/8.4
 $ # <intense conflict resolution>
 $ git commit
 $ git merge origin/improvement/CLDSRV-167-put-bucket-tagging
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.4/improvement/CLDSRV-167-put-bucket-tagging
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

KillianG commented:
/approve

Commit:
CLDSRV-1167: use arsenal parse tag function

bert-e commented:
# Conflict




A conflict has been raised during the update of
integration branch `w/8.4/improvement/CLDSRV-167-put-bucket-tagging` with contents from `improvement/CLDSRV-167-put-bucket-tagging`
and `development/8.4`.

Please resolve the conflict on **the integration branch** (`w/8.4/improvement/CLDSRV-167-put-bucket-tagging`).


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout w/8.4/improvement/CLDSRV-167-put-bucket-tagging
 $ git pull  # or "git reset --hard origin/w/8.4/improvement/CLDSRV-167-put-bucket-tagging"
 $ git merge origin/development/8.4
 $ # <intense conflict resolution>
 $ git commit
 $ git merge origin/improvement/CLDSRV-167-put-bucket-tagging
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.4/improvement/CLDSRV-167-put-bucket-tagging
```

KillianG commented:
/approve

bert-e commented:
# Conflict




A conflict has been raised during the update of
integration branch `w/8.4/improvement/CLDSRV-167-put-bucket-tagging` with contents from `improvement/CLDSRV-167-put-bucket-tagging`
and `development/8.4`.

Please resolve the conflict on **the integration branch** (`w/8.4/improvement/CLDSRV-167-put-bucket-tagging`).


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout w/8.4/improvement/CLDSRV-167-put-bucket-tagging
 $ git pull  # or "git reset --hard origin/w/8.4/improvement/CLDSRV-167-put-bucket-tagging"
 $ git merge origin/development/8.4
 $ # <intense conflict resolution>
 $ git commit
 $ git merge origin/improvement/CLDSRV-167-put-bucket-tagging
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.4/improvement/CLDSRV-167-put-bucket-tagging
```





*The following options are set:* **approve**

KillianG commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/7.10`

* :heavy_check_mark: `development/8.4`

* :heavy_check_mark: `development/8.5`



The following branches will **NOT** be impacted:

* `development/6.4`
* `development/7.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/7.10`

* :heavy_check_mark: `development/8.4`

* :heavy_check_mark: `development/8.5`



The following branches have **NOT** changed:

* `development/6.4`
* `development/7.4`



Please check the status of the associated issue CLDSRV-167.

Goodbye killiang.
